### DISCUSSION TITLE
E2E tests and new PRs

### DISCUSSION THREAD

**Discussion by gigitux** [woocommerce/woocommerce#44951] [WooCommerce, Blocks, and Site Editing]
I'd like to kick off a discussion regarding our approach to new pull requests and whether we should mandate the inclusion of E2E tests before merging them into our codebase. **This is an important decision that can significantly impact our code quality and reliability.**

### Background
As our project has grown and more folks are involved in the project, we can see how fast we are shipping new enhancements, but at the same time, the complexity of our codebase is increasing a lot. During the last months, we have seen an increase in issues related to regressions and unintended side effects of new features or changes. E2E tests can play a pivotal role in identifying these issues early in the development process, preventing them from reaching the production environment.

We recognized that our previous E2E infrastructure had its share of reliability issues, and the developer experience left much to be desired. To address these concerns, over the past few months, we have dedicated a significant amount of time and effort to revamp and enhance the E2E infrastructure, as documented in issue #9001 (https://github.com/woocommerce/woocommerce-blocks/issues/9001).

While we still have other improvements in the pipeline, such as adding support for multiple versions of WordPress/Gutenberg, the newly restructured E2E infrastructure appears to be much more stable and with a good DX.

I think that it is time to iterate on our test culture, starting to decide when a PR should include E2E tests.

### Some Points to Consider:

**Challenges and Costs**: It's important to acknowledge that implementing E2E tests requires additional time and effort. What are the potential challenges we might face, and how can we mitigate them? Are there specific areas of our application where E2E tests are particularly valuable?

**PR Review Process:** Should we make E2E tests a mandatory part of the PR review checklist? What criteria should we use to determine whether E2E tests are necessary for a particular PR?

**Exceptions and Guidelines**: Are there any situations or types of PRs where E2E tests might not be required? For example, should small bug fixes also include E2E tests?

**Education**: If we decide to move forward with requiring E2E tests, how can we ensure that contributors are proficient in writing and maintaining them?

What do you think? It's essential that we make an informed decision that aligns with our project goals and development standards.

Kudos to @opr for reviewing this post!

**Comment by nerrad** [comment]
In general, I agree about the necessity of adding tests on every PR (and even mandating that). However, I think mandating a specific _type_ of test should be off the table as a consideration. E2E tests have their value but I don't think every PR would (or should) require the overhead of E2E.

E2E tests are the most time-consuming and complex type of test to implement and run and should be reserved for testing complete user flows.

Some code changes would be better tested via integration or unit tests and those should always be considered before jumping immediately to E2E. Within WordPress development, pure unit tests are rarer so we typically implement integration more than unit tests, regardless, there are many occasions where an integration/unit test is _more_ valuable than an E2E test because it's faster to run, less complex to write/understand, and more resilient against flakiness (or false fails). Helper functions, components, and hooks are all good candidates for unit/integration tests.

I'm definitely 💯 on requiring automated test coverage on PRs. However, the type of tests should be geared towards the changes being implemented (rather than mandating one type of test). 

It's also tricky to prescribe a formula for when to choose a specific type of test for a given PR given the content of the PR may have changes that are best tested via multiple types. This incidentally, is one of the reasons why keeping PRs smaller is a significant help not only for the reviewer and time to merge but also for understanding what type of test to write.

## To illustrate - Imagine this flow.

Approach A:
- Code is written for a large new block that includes some helper functions, components, and the block itself. 
- The code author decides the PR is already large so writes just an E2E test to cover the behaviour of the block.
- The E2E tests pass (the user behaviour is preserved!) the PR is reviewed, and the code is merged into trunk and released.

Approach B:
- Code is written for a large new block that includes some helper functions, components, and the block itself.
- The code author decides this is a really large changeset, so breaks it up into smaller PRs. One for each of the public helper functions and components, one for the new block using all of those pieces.
- In each of the smaller public helper and component PRs, the code author writes unit/integration tests to both validate their behaviour and also describe how they behave (the public "contract"). These are reviewed and merged.
- In the block PR, the code author writes a E2E test to cover just the user behaviour of interacting with the block in context of the rest of the app. This is reviewed and merged after the other PRs.

Now let's think about the resiliency of the above approaches keeping in mind, we've got some helper functions and components that are not internal and available for other usecases and we have the block. 

The scenario is that sometime after the release of the block, a change is made in one of the components or helper functions which introduces a bug. What _could_ happen with either of the above approaches?

Approach A:

- The E2E test for the block using that helper function might fail. **Or it might not** if it's not covering a user flow in a way that surfaces the bug. 
- Let's say the E2E test does fail. The developer discovers which block behaviour is covered by the E2E test and naively thinks because this is a problem in the block, addresses the problem _in the block_. So the developer happily works_ around the _bug_ actually happening in the function/component by doing the fix in the *block*. The Block E2E test passes and the PR is merged and everything is roses again. But is it?

Approach B:
- The test on the function/component where the bug is introduced fails. The E2E test might not fail, but it doesn't matter at that point because the integration/unit test has caught where the actual problem is.
- The bug is fixed, the unit/integration tests pass (and if the E2E test failed it should pass too).
- The PR is merged. 

Which approach seems better?

**Reply by gigitux** [reply]
Thank you, @nerrad - I completely agree with your perspective, and I genuinely appreciate your input. Upon revisiting the post I wrote, I realized that it may be E2E-focused. The primary objective of this post, however, is to establish a broader framework for finding a robust testing culture within our development process.

> I'm definitely 💯 on requiring automated test coverage on PRs. However, the type of tests should be geared towards the changes being implemented (rather than mandating one type of test).

Yeah, for example, for https://github.com/woocommerce/woocommerce-blocks/pull/10999 I used a TDD approach. In this way, I had a fast feedback loop checking the progress of the algorithm I wrote.

I'm a big fan of small PRs. This is one of my favorite resources about small PRs: https://google.github.io/eng-practices/review/developer/small-cls.html. Also, this meme is very famous:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/d2f93a1d-d796-4315-aa3c-f80b2ad5daa8)


> Which approach seems better?

Approach B is better! 


On the other side, I suspect it is very complex to write integration tests for the blocks (especially on the editor side), since we should mock a lot of Gutenberg logic.

From my experience (and looking at Gutenberg's approach), I would say that:

- utils/hooks and pure function could be easily tested with unit test/integrations test.
- blocks can be easily tested with E2E tests.


It isn't possible to set "a standard," but as a team, during the review, we should be stricter and don't approve PR if we don't have any coverage, especially if:
- we are adding a new feature.
- we are fixing a bug that would be hard to catch during the release testing.
